### DISCUSSION TITLE
build/deb/build_deb.sh script failed with brz: ERROR: unknown command "dh-make"

### DISCUSSION THREAD

**Discussion by liorf95** [milvus-io/milvus#33735] [Q&A and General discussion]
I am trying to build the Debian package from Milvus source code.
However the build/deb/build_deb.sh script failed with brz: ERROR: unknown command "dh-make".

Is it some dependency version issue?
Any idea?

**Comment by xiaofan-luan** [comment]
i'm afraid debian compilation is out of maintenance for a while.

Our recommendation is ubuntu 22.04, rock linux or aws linux

@zwd1208 do we have an environment running on debian?

**Reply by liorf95** [reply]
The reported issue is on ubuntu 22.04.
On ubuntu 20.04 I am able to downgrade brz-debian and bzr dh-make do working.

However now bzr builddeb -- -us -uc failed due to "dh: error: Please specify the compatibility level in debian/compat" although the debian/compat file is 10 as it should.

**Comment by xiaofan-luan** [comment]
anyone would like to help?

**Comment by xiaofan-luan** [comment]
build/deb/build_deb.sh:apt install gnupg pbuilder ubuntu-dev-tools apt-file dh-make build-essential -y
build/deb/build_deb.sh:bzr dh-make milvus $2 milvus-$2.tar.gz

dh-make should be installed automatically with build_deb.sh.

We are running on ubuntu 22.04 as well for the ci/cd env and it seems to be working perfect

**Reply by liorf95** [reply]
1) Do you run the build_deb.sh script as root?
2) From what path location do you run the build_deb.sh script?

**Reply by xiaofan-luan** [reply]
guess you can run the scripts on any where becasue it is relative path.
you can run apt install see if you can install correctly

**Reply by liorf95** [reply]
One more question please:
The command bzr dh-make milvus $2 milvus-$2.tar.gz in line 46 is interactive - Is there a way to automate it? - 
The --single --yes flags are not recognized by the bzr.
Also, is there a way to speed up the dh-make operation? It is very slow (more than an hour!!)

**Comment by tespio** [comment]
Sorry for reviving a (somewhat old) thread, but is anyone still trying to build .deb packages for milvus 2.4.x ? I tried myself, on Debian 12 and Ubuntu 22.04 (as well as Ubuntu 20.04) and I fail building the deb with both docker build method or manually. Latest .deb build on release seems to be on 2.3.x but I would like to also test the 2.4 and I wanted to buld the .deb package for it instead of running it from sources. Guess if nobody takes a look at the conflicts between breezy and dh-make and can make the deb build use debuild of something else instead of dh-make.

**Reply by liorf95** [reply]
We are building deb pkg for 2.4.x using the scripts here with some of our adjustments - what issue are facing?

**Reply by tespio** [reply]
I managed to build it. But specifically by installing only brz-debian_2.8.42_all.deb (this particular version) and then doing
```shell
apt-mark hold brz-debian
``` 
Even like this i still get the error below, but it goes on and finishes properly without upgrading it even if it says so:
```shell
The following packages will be upgraded:
  brz-debian
1 upgraded, 0 newly installed, 0 to remove and 16 not upgraded.
E: Held packages were changed and -y was used without --allow-change-held-packages.
```

**Comment by liorf95** [comment]
Why don't you use the Debian package build procedure below to avoid all the brz-debian issues?
https://github.com/milvus-io/milvus/tree/master/build/deb

**Reply by tespio** [reply]
I did exactly that. But the docker image that gets downloaded when you run docker and the build script, will update the brz-debian to the latest version that doesn't work with dh_make. Maybe you are still using an older docker image that comes with brz-debian 2.8.42 and it works out of the box.

**Reply by liorf95** [reply]
It could be an Ubuntu platform issue - it always installs brz-debian 2.8.32 version when we run it.
We are using Ubuntu 22.04.3 LTS.
More importantly - we always take the milvus:v2.4.6 release docker image.
docker run -d --name build-deb -v .:/make_deb -ti --entrypoint /bin/bash milvusdb/milvus:v2.4.6
